aspects.bzl

```diff
@@ -0,0 +1,228 @@
+# Copyright 2020 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Rust Analyzer Bazel rules.
+
+rust_analyzer will generate a rust-project.json file for the
+given targets. This file can be consumed by rust-analyzer as an alternative
+to Cargo.toml files.
+"""
+
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+_rust_rules = [
+    "rust_library",
+    "rust_binary",
+]
+
+TargetInfo = provider(
+    fields = {
+        'name' : 'target name',
+        'root' : 'crate root',
+        'edition' : 'edition',
+        'dependencies' : 'dependencies',
+        'transitive_deps' : 'closure of all transitive dependencies',
+        'cfgs' : 'compilation cfgs',
+        'env' : 'Environment variables, used for the `env!` macro',
+    }
+)
+
+def _rust_project_aspect_impl(target, ctx):
+  # We support only these rule kinds.
+  if ctx.rule.kind not in _rust_rules:
+    return []
+
+  info = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"]
+
+  # extract the crate_root path
+  edition = ctx.rule.attr.edition
+  if not edition:
+    # TODO if edition isn't specified, default to what is in the rust toolchain
+    edition = info.default_edition
+
+  crate_name = ctx.rule.attr.name
+
+  crate_root = ctx.rule.attr.crate_root
+  if not crate_root:
+    if len(ctx.rule.attr.srcs) == 1:
+      crate_root = ctx.rule.attr.srcs[0]
+    else:
+      for src in ctx.rule.attr.srcs:
+        file_name = src.label.name
+        crate_type = ctx.rule.attr.crate_type
+        if crate_type == "bin" and file_name.endswith("main.rs"):
+          crate_root = src
+          break
+        if crate_type == "rlib" or crate_type == "dylib":
+          if file_name.endswith("lib.rs"):
+            crate_root = src
+            break
```

Comment:
I wonder if we have this logic somewhere else, I think it would be better to extract it in a utility method.